Replace visual character and allow for replacement by modules

flo opened the pull request:
### Contains

This PR:
* This PR introduces a new event (CreateVisualCharacterEvent) that modules can handle to create their own character visualization 
  * The visualization is completely up to the module. It could be a static model, a animated one or theoreticlly just a particle system.
  * The model can but does not need to be colored like the player. It is up to the event handler
  * For the own character no CreateVisualCharacterEvent gets fired so that "hideFromOwner" flag is needed for the SkeletalMeshComponent 
  * If the event does not get handled, a floating cube will be shown as character model (instead of the moneky head)
* The existing monkey head has been removed as it has been replaced by the floating cube.

The reason for replacing the default monkey head with a default floating cube is that the monkey head is not so clearly a place holder and people would just wonder why there is a monkey head. Also it is better to show of an animated model instead of a static one.

### How to test
Unit tests:
* Run VisualCharacterSystemTest

Manual testing:
1. Start the game as listen server.
2. Start another instance of the game and connect to server
3. Expected result: The own character is invisible in each instance while the other is a cube floating slightly and down constantly.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

Josharias commented on engine/src/main/java/org/terasology/logic/characters/VisualCharacterComponent.java:
```diff
@@ -0,0 +1,32 @@
+/**
+ * Add this component to characters to give them a visual appearance.
+ *
+ * This component does not specifcy however how the character look like. Instead
```
specifcy == "specify"

Josharias commented on engine/src/main/java/org/terasology/logic/characters/VisualCharacterSystem.java:
```diff
@@ -0,0 +1,128 @@
+        sendCreateVisualCharacterEventIfNotOwnCharacter(entity);
+    }
+
+    void sendCreateVisualCharacterEventIfNotOwnCharacter(EntityRef entity) {
```
I wonder,  if in some future land we have a 3rd person camera,  we may likely want to have this event still sent. Maybe with a flag that it is for the local player,  and then let the mesh adding system do the logic whether or not it wants to add it for the local player's character?

Alternatively,  cancel the event if the player is in first person mode?  (not sure if this would be a good idea)

Josharias commented on engine/src/main/java/org/terasology/logic/characters/VisualCharacterSystem.java:
```diff
@@ -0,0 +1,128 @@
+     * The CreateVisualCharacterEvent events that could not be sent previously will be sent.
+     */
+    @ReceiveEvent
+    public void onAwaitedLocalCharacterSpawnEvent(AwaitedLocalCharacterSpawnEvent event, EntityRef entity) {
```
I am curious,  what happens if you add meshes and such before spawn?

Josharias commented:
Also, changing color didnt seem to affect the mesh color on rejoining a game. My chat color would change,  but not the mesh color.

emanuele3d commented:
I just wanted to say I did see this PR and plan to review it! Bear with me!

Cervator commented:
Bump, particularly for @flo :-) Would like to see this finishing up soon so we can start looking at additional characters.

flo commented:
Ah didn't test it with 3 instances, just with 2. I added the cleanup of the visual character entity.

Also I added some more explanation.

An open issue is still the players color: Rejoining players are not properly colored as the character gets loaded (and thus the VisualComponent activated) before the character got linked to the player. So once you rejoin a game your character is white. I will have to think about how to address it.

flo commented:
@Josharias  and thanks for the review btw

GooeyHub commented:
Uh oh, something went wrong with the build. Need to check on that

Cervator commented:
Thanks for the updates :-)

Build problem seems minor, a single unit test broke: http://jenkins.terasology.org/job/TerasologyPRs/2094/testReport/junit/org.terasology.logic.characters/VisualCharacterSystemTest/testSendingOfCreateVisualCharacterEvent/

flo commented:
Module for testing it out: https://github.com/Terasology/HumanoidCharacters

flo commented:
When testing it out with character module please not that the character is wrongly turned.

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

emanuele3d commented on engine/src/main/resources/assets/materials/floatingCubeSkin.mat:
```diff
@@ -0,0 +1,12 @@
+    ],
+    "textured": true
+  }
+}
```
Something funny with the EOF here. Add a new line perhaps?

flo commented:
@emanuele3d thanks for review, I worked it in. I also changed the map to a field in the component to allow systems to get from character to visual character.

GooeyHub commented:
Uh oh, something went wrong with the build. Need to check on that

GooeyHub commented:
Uh oh, something went wrong with the build. Need to check on that

GooeyHub commented:
Uh oh, something went wrong with the build. Need to check on that

Cervator commented on engine-tests/src/test/java/org/terasology/logic/characters/VisualCharacterSystemTest.java:
```diff
@@ -0,0 +1,183 @@
+         * Simulate activation before entity is done
+         * since the character is not properly linked yet nothing should happen
+         */
+        system.onActivatedVisualCharacter(OnActivatedComponent.newInstance(), otherCharacterEntity);
```
Syntax error here and two further instance of the same situation - no `VisualCharacterComponent` is supplied to the method despite being required. I suspect the target method was changed after the test was written? Easy to miss since `engine-tests` won't force recompile when just trying to re-run the game for play testing.

This is what is marking the Jenkins build as broken. I tried just throwing in a `new VisualCharacterComponent()` in the three spots but then the test hit an similar sounding NPE so it probably needs a quick bit of tweaking

Cervator commented on engine-tests/src/test/java/org/terasology/logic/characters/VisualCharacterSystemTest.java:
```diff
@@ -0,0 +1,183 @@
+     *     <li>No visual character gets created (via event) for the own character (as it is first person)</li>
+     *     <li>That the system can deal with LocalPlayer and characters not being properly linked when
+     *     the character entities get loaded/created</li>
+     *     <li>A visual character gets created (via event) for charcters that were already present when the player
```
Typo: "charcters"

Cervator commented on engine-tests/src/test/java/org/terasology/logic/characters/VisualCharacterSystemTest.java:
```diff
@@ -0,0 +1,183 @@
+/*
+ * Copyright 2017 MovingBlocks
```
2018 :D

Cervator commented on engine-tests/src/test/java/org/terasology/logic/characters/VisualCharacterSystemTest.java:
```diff
@@ -0,0 +1,183 @@
+
+
+        /*
+         * There is no second AwaitedLocalCharacterSpawnEvent event,
```
This is probably short enough to just be a regular single line comment?

Cervator commented on engine/src/main/java/org/terasology/logic/characters/VisualCharacterComponent.java:
```diff
@@ -0,0 +1,41 @@
+/**
+ * Add this component to characters to give them a visual appearance.
+ *
+ * This component does not specify however how the character look like.
```
Lines could probably be a bit longer + excess blank lines

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

GooeyHub commented:
Hooray Jenkins reported success with all tests good!

J-Dill commented:
I was able to fix the Null Pointer Exception in the test in my commit that I referenced above. I am not sure on how to get the changes to you, however, @flo. How do I go about doing that?

flo commented:
@J-Dill oh, that sounds like the changes I made. Sorry that this turned out in redundant work but I didn't expect someone to do a PR for a small PR like this.

J-Dill commented:
Ahh yeah, no problem! For future reference though, is that what I am supposed to do when suggesting a fix: create my own fix and then reference this PR when I merge into my own? Did that work alright? @flo

flo commented:
I would suggest to just ping the owner of the PR first and ask if you can help out. That way double work is avoided. Otherwise you can create a PR on my repository for merging your branch into mine.

J-Dill commented:
I see, thanks!
